File: tests/test_llm_advisor.py
"""
LLM Advisor — Optional advisory analysis tests.
Python equivalent of test-llm-advisor.mjs
Run: python3 -m unittest tests/test_llm_advisor.py -v
"""

import os
import sys
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "python"))
from quadruple_verification import _parse_advisory_findings, _run_llm_advisory


# ─── _parse_advisory_findings ─────────────────────────────────────────────────

class TestParseAdvisoryFindings(unittest.TestCase):
    def test_parses_valid_json_array(self):
        text = '[{"severity":"high","category":"security","description":"SQL injection","line_hint":"15"}]'
        result = _parse_advisory_findings(text)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0]["severity"], "high")
        self.assertEqual(result[0]["category"], "security")

    def test_parses_empty_array(self):
        self.assertEqual(_parse_advisory_findings("[]"), [])

    def test_parses_json_in_code_block(self):
        text = '```json\n[{"severity":"low","category":"quality","description":"test","line_hint":"1"}]\n```'
        result = _parse_advisory_findings(text)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0]["severity"], "low")

    def test_extracts_array_from_surrounding_text(self):
        text = (
            "Here are the findings:\n"
            '[{"severity":"medium","category":"performance","description":"slow loop","line_hint":"42"}]\n'
            "End of analysis."
        )
        result = _parse_advisory_findings(text)
        self.assertEqual(len(result), 1)
        self.assertEqual(result[0]["category"], "performance")

    def test_returns_empty_for_none(self):
        self.assertEqual(_parse_advisory_findings(None), [])
        self.assertEqual(_parse_advisory_findings(""), [])

    def test_returns_empty_for_invalid_json(self):
        self.assertEqual(_parse_advisory_findings("not json at all"), [])
        self.assertEqual(_parse_advisory_findings("{invalid}"), [])

    def test_handles_multiple_findings(self):
        import json
        findings = [
            {"severity": "high", "category": "security", "description": "XSS", "line_hint": "10"},
            {"severity": "medium", "category": "quality", "description": "Missing null check", "line_hint": "25"},
            {"severity": "low", "category": "performance", "description": "Unnecessary loop", "line_hint": "40"},
        ]
        result = _parse_advisory_findings(json.dumps(findings))
        self.assertEqual(len(result), 3)


# ─── _run_llm_advisory guard conditions ───────────────────────────────────────

class TestRunLlmAdvisoryGuards(unittest.TestCase):
    def test_returns_empty_when_disabled(self):
        result = _run_llm_advisory("some content", {"llm": {"enabled": False}})
        self.assertEqual(result, [])

    def test_returns_empty_when_no_llm_key(self):
        result = _run_llm_advisory("some content", {})
        self.assertEqual(result, [])

    def test_returns_empty_when_content_too_short(self):
        result = _run_llm_advisory("short", {"llm": {"enabled": True}})
        self.assertEqual(result, [])

    def test_returns_empty_when_content_empty(self):
        result = _run_llm_advisory("", {"llm": {"enabled": True}})
        self.assertEqual(result, [])

    def test_returns_empty_when_api_key_not_set(self):
        original = os.environ.pop("ANTHROPIC_API_KEY", None)
        try:
            result = _run_llm_advisory("a" * 100, {"llm": {"enabled": True}})
            self.assertEqual(result, [])
        finally:
            if original is not None:
                os.environ["ANTHROPIC_API_KEY"] = original


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: tests/test-llm-advisor.mjs
import { describe, it } from 'node:test';
import { strict as assert } from 'node:assert';
import { parseFindings, ADVISORY_PROMPT, runAdvisory } from '../scripts/lib/llm-advisor.mjs';

// ─── parseFindings ───────────────────────────────────────────────────────────

describe('LLM Advisor — parseFindings()', () => {
  it('parses a valid JSON array', () => {
    const input = '[{"severity":"high","category":"security","description":"SQL injection","line_hint":"15"}]';
    const result = parseFindings(input);
    assert.equal(result.length, 1);
    assert.equal(result[0].severity, 'high');
    assert.equal(result[0].category, 'security');
  });

  it('parses empty array', () => {
    const result = parseFindings('[]');
    assert.deepEqual(result, []);
  });

  it('parses JSON in code block', () => {
    const input = '```json\n[{"severity":"low","category":"quality","description":"test","line_hint":"1"}]\n```';
    const result = parseFindings(input);
    assert.equal(result.length, 1);
    assert.equal(result[0].severity, 'low');
  });

  it('extracts array from surrounding text', () => {
    const input = 'Here are the findings:\n[{"severity":"medium","category":"performance","description":"slow loop","line_hint":"42"}]\nEnd of analysis.';
    const result = parseFindings(input);
    assert.equal(result.length, 1);
    assert.equal(result[0].category, 'performance');
  });

  it('returns empty array for null/undefined', () => {
    assert.deepEqual(parseFindings(null), []);
    assert.deepEqual(parseFindings(undefined), []);
    assert.deepEqual(parseFindings(''), []);
  });

  it('returns empty array for invalid JSON', () => {
    assert.deepEqual(parseFindings('not json at all'), []);
    assert.deepEqual(parseFindings('{invalid}'), []);
  });

  it('handles multiple findings', () => {
    const input = JSON.stringify([
      { severity: 'high', category: 'security', description: 'XSS', line_hint: '10' },
      { severity: 'medium', category: 'quality', description: 'Missing null check', line_hint: '25' },
      { severity: 'low', category: 'performance', description: 'Unnecessary loop', line_hint: '40' }
    ]);
    const result = parseFindings(input);
    assert.equal(result.length, 3);
  });
});

// ─── ADVISORY_PROMPT ─────────────────────────────────────────────────────────

describe('LLM Advisor — ADVISORY_PROMPT', () => {
  it('instructs to check security vulnerabilities', () => {
    assert.ok(ADVISORY_PROMPT.includes('Security vulnerabilities'));
  });

  it('instructs to check logic errors', () => {
    assert.ok(ADVISORY_PROMPT.includes('Logic errors'));
  });

  it('instructs to check performance', () => {
    assert.ok(ADVISORY_PROMPT.includes('Performance concerns'));
  });

  it('instructs to return JSON array', () => {
    assert.ok(ADVISORY_PROMPT.includes('Return a JSON array'));
    assert.ok(ADVISORY_PROMPT.includes('Return ONLY the JSON array'));
  });
});

// ─── runAdvisory guard conditions ────────────────────────────────────────────

describe('LLM Advisor — runAdvisory() guards', () => {
  it('returns empty when llm.enabled is false', async () => {
    const result = await runAdvisory('some content', { llm: { enabled: false } });
    assert.deepEqual(result, []);
  });

  it('returns empty when config has no llm key', async () => {
    const result = await runAdvisory('some content', {});
    assert.deepEqual(result, []);
  });

  it('returns empty when content is too short', async () => {
    const result = await runAdvisory('short', { llm: { enabled: true } });
    assert.deepEqual(result, []);
  });

  it('returns empty when content is empty', async () => {
    const result = await runAdvisory('', { llm: { enabled: true } });
    assert.deepEqual(result, []);
  });

  it('returns empty when ANTHROPIC_API_KEY is not set', async () => {
    const originalKey = process.env.ANTHROPIC_API_KEY;
    delete process.env.ANTHROPIC_API_KEY;
    try {
      const result = await runAdvisory('a'.repeat(100), { llm: { enabled: true } });
      assert.deepEqual(result, []);
    } finally {
      if (originalKey) process.env.ANTHROPIC_API_KEY = originalKey;
    }
  });
});
